hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+      throws InterruptedException, ExecutionException {
+    CompletableFuture<Void> future = setHDFSAclParallel(hdfsAclOperations.getSecond());
+    for (HDFSAclOperation hdfsAclOperation : hdfsAclOperations.getFirst()) {
+      setSingleHDFSAcl(hdfsAclOperation).get();
+    }
+    future.get();
+  }
+
+  private CompletableFuture<Void> setSingleHDFSAcl(HDFSAclOperation acl) {
+    return CompletableFuture.supplyAsync(() -> {
+      try {
+        acl.setAcl();
+      } catch (IOException e) {
+        LOG.error("Set HDFS acl error for path {}", acl.path.toString(), e);
+      }
+      return acl;
+    }, pool).thenComposeAsync(acl2 -> {
+      List<HDFSAclOperation> childAclOperations = null;
+      try {
+        childAclOperations = acl2.getChildAclOperations();
+      } catch (IOException e) {
+        LOG.error("Set HDFS acl error for path {}", acl2.path.toString(), e);
+      }
+      return setHDFSAclParallel(childAclOperations);
+    }, pool);
+  }
+
+  private CompletableFuture<Void> setHDFSAclParallel(List<HDFSAclOperation> operations) {
+    List<CompletableFuture<Void>> futures = operations.stream()
+        .map(operation -> setSingleHDFSAcl(operation)).collect(Collectors.toList());
+    CompletableFuture<Void> future =
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+    return future;
+  }
+
+  private static AclEntry aclEntry(AclEntryScope scope, String name, FsAction action) {
+    return new AclEntry.Builder().setScope(scope)
+        .setType(AuthUtil.isGroupPrincipal(name) ? GROUP : USER).setName(name).setPermission(action)
+        .build();
+  }
+
+  /**
+   * Inner class used to describe modify or remove what acl entries for files or directories(and
+   * child files)
+   */
+  private static class HDFSAclOperation {
+    enum OperationType {
+      MODIFY, REMOVE
+    }
+
+    interface Operation {
+      void apply(FileSystem fs, Path path, List<AclEntry> aclList) throws IOException;
+    }
+
+    private FileSystem fs;
+    private Path path;
+    private List<AclEntry> dirAcl;
+    private List<AclEntry> fileAcl;
+    private boolean recursive;
+    private Operation operation;
+    private List<Path> skipPaths;
+
+    HDFSAclOperation(FileSystem fs, Path path, Set<String> users, OperationType operationType,
+        FsAction fsAction, boolean recursive, List<Path> skipPaths) {
+      this.fs = fs;
+      this.path = path;
+      this.dirAcl = getDefaultAclEntries(users, fsAction);
+      this.fileAcl = getAccessAclEntries(users, fsAction);
+      this.recursive = recursive;
+      this.skipPaths = skipPaths;
+      if (operationType == OperationType.MODIFY) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.modifyAclEntries(modifyPath, aclList);
+      } else if (operationType == OperationType.REMOVE) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.removeAclEntries(modifyPath, aclList);
+      }
```

Comment:
If operationType is unexpected,  then should throw an IllegalArgumentException ?

Reply:
done